Add doc for indexer property (#2860)
dotnet/EntityFramework.Docs: entity-framework/core/modeling/shadow-properties.md

---
# Shadow and Indexer Properties

Shadow properties are properties that aren't defined in your .NET entity class but are defined for that entity type in the EF Core model. The value and state of these properties is maintained purely in the Change Tracker. Shadow properties are useful when there's data in the database that shouldn't be exposed on the mapped entity types.

> Indexer properties are entity type properties, which are backed by [indexer](/dotnet/csharp/programming-guide/indexers/) in .NET entity class. They can be accessed using indexer on the .NET class instances. It also allows you to add additional properties in the entity type model without changing the CLR class. If the CLR type has indexer defined, then when creating shadow properties conventionally, EF Core by default creates an indexer property.

Reviewer: ```suggestion
Indexer properties are entity type properties, which are backed by an [indexer](/dotnet/csharp/programming-guide/indexers/) in .NET entity class. They can be accessed using the indexer on the .NET class instances. It also allows you to add additional properties to the entity type without changing the CLR class.
```

Shadow properties cannot be accessed after a no-tracking query since the entities returned are not tracked by the change tracker.

## Configuring indexer properties

> You can use the Fluent API to configure indexer properties. Once you've called the string `IndexerProperty`, you can chain any of the configuration calls you would for other properties. In the following sample, `Blog` has indexer defined and it will be used to create indexer property.

Reviewer: ```suggestion
You can use the Fluent API to configure indexer properties. Once you've called the method `IndexerProperty`, you can chain any of the configuration calls you would for other properties. In the following sample, `Blog` has an indexer defined and it will be used to create an indexer property.
```

[!code-csharp[Main](../../../samples/core/Modeling/FluentAPI/IndexerProperty.cs?name=ShadowProperty&highlight=3)]

> If the name supplied to the `IndexerProperty` method matches the name of an existing indexer property, then the code will configure that existing property. If the entity type has a property, which is backed by a property on the entity class, then an exception is thrown since indexer properties must be accessed via indexer only.

Reviewer: ```suggestion
If the name supplied to the `IndexerProperty` method matches the name of an existing indexer property, then the code will configure that existing property. If the entity type has a property, which is backed by a property on the entity class, then an exception is thrown since indexer properties must only be accessed via the indexer.
```

## Property bag entity types

> [!NOTE]
> Support for Property bag entity types was added in EF Core 5.0.

> Entity types that contain only indexer properties are known as property bag entity types. These entity types don't have shadow properties. Currently only `Dictionary<string, object>` is supported as a property bag entity type. It must be configured as a shared entity type with a unique name and the corresponding `DbSet` property must be implemented using a `Set` call.

Reviewer: ```suggestion
Entity types that contain only indexer properties are known as property bag entity types. These entity types don't have shadow properties, instead EF will create indexer properties. Currently only `Dictionary<string, object>` is supported as a property bag entity type. It must be configured as a shared entity type with a unique name and the corresponding `DbSet` property must be implemented using a `Set` call.
```
